syntax = "proto3";

import "google/protobuf/timestamp.proto";

service Chat {
    rpc GetUserChats (GetUserChatsRequest) returns (MultipleChatsResponse) {}
    rpc CreateChat (CreateChatRequest) returns (ChatResponse) {}
    rpc AddMembersToChat (AddMembersToChatRequest) returns (ChatResponse) {}
    rpc GetChatData (GetChatDataRequest) returns (ChatData) {}
    rpc DeleteUserChat (DeleteUserChatRequest) returns (DeleteResponse) {}
    rpc DeleteChat (DeleteChatRequest) returns (DeleteResponse) {}

}

message ChatMember {
    int32 id = 1;
}

message GetUserChatsRequest {
    int32 user_id = 1;
}

message ChatResponse {
    int32 id = 1;
    string name = 2;
    optional string avatar = 3;
    optional string last_message = 4;
    optional google.protobuf.Timestamp last_message_at = 5;
}

message MultipleChatsResponse {
    repeated ChatResponse chats = 1;
}

message CreateChatRequest {
    string name = 1;
    repeated ChatMember members = 2;
}

message AddMembersToChatRequest {
    int32 chat_id = 1;
    repeated ChatMember members = 2;
}

message GetChatDataRequest {
    int32 chat_id = 1;
}

message FullChatMember {
    int32 id = 1;
}

message ChatData {
    int32 id = 1;
    string name = 2;
    optional string avatar = 3;
    optional string last_message = 4;
    optional google.protobuf.Timestamp last_message_at = 5;
    repeated FullChatMember members = 6;
    string created_at = 7;
}

message DeleteUserChatRequest {
    int32 user_id = 1;
    int32 chat_id = 2;
}

message DeleteChatRequest {
    int32 chat_id = 1;
}

message DeleteResponse {
    string status = 1;
}
